staging/src/k8s.io/apiserver/pkg/endpoints/handlers/patch.go

```diff
@@ -446,28 +449,97 @@ func (p *patcher) applyAdmission(ctx context.Context, patchedObject runtime.Obje
+		newObjVersioned, err := convert(newObj)
+		if err != nil {
+			return nil, fmt.Errorf("failed to convert new object to proper version: %v", err)
+		}
+		oldObjVersioned, err := convert(oldObj)
+		if err != nil {
+			return nil, fmt.Errorf("failed to convert old object to proper version: %v", err)
+		}
+
+		managed, err := apply.DecodeObjectManagedFields(newObj)
+		if err != nil {
+			return nil, err
+		}
+
+		newObjTyped, err := typeConverter.ObjectToTyped(newObjVersioned)
+		if err != nil {
+			return nil, fmt.Errorf("failed to create typed new object: %v", err)
+		}
+		oldObjTyped, err := typeConverter.ObjectToTyped(oldObjVersioned)
+		if err != nil {
+			return nil, fmt.Errorf("failed to create typed old object: %v", err)
+		}
+
+		updater := merge.Updater{Converter: failConverter{}}
+		managed, err = updater.Update(oldObjTyped, newObjTyped, managed, manager)
+		if err != nil {
+			return nil, fmt.Errorf("failed to update ManagedFields: %v", err)
+		}
+
+		if err := apply.EncodeObjectManagedFields(newObj, managed); err != nil {
+			return nil, err
+		}
+
+		return newObj, nil
+	}
+}
+
 // patchResource divides PatchResource for easier unit testing
 func (p *patcher) patchResource(ctx context.Context, scope RequestScope) (runtime.Object, bool, error) {
+	converter, err := apply.NewTypeConverter(scope.OpenAPIModels)
+	if err != nil {
+		return nil, false, err
+	}
 	p.namespace = request.NamespaceValue(ctx)
 	switch p.patchType {
 	case types.JSONPatchType, types.MergePatchType:
 		p.mechanism = &jsonPatcher{patcher: p}
+		p.applyManagedFields = makeManagedFieldsUpdater(converter, newConvertor(p.unsafeConvertor, p.kind.GroupVersion()), "jsonpatch")
 	case types.StrategicMergePatchType:
 		schemaReferenceObj, err := p.unsafeConvertor.ConvertToVersion(p.restPatcher.New(), p.kind.GroupVersion())
 		if err != nil {
 			return nil, false, err
 		}
 		p.mechanism = &smpPatcher{patcher: p, schemaReferenceObj: schemaReferenceObj}
+		p.applyManagedFields = makeManagedFieldsUpdater(converter, newConvertor(p.unsafeConvertor, p.kind.GroupVersion()), "smpatch")
 	// this case is unreachable if ServerSideApply is not enabled because we will have already rejected the content type
 	case types.ApplyPatchType:
-		p.mechanism = &applyPatcher{patcher: p, model: scope.OpenAPISchema}
+		updater := merge.Updater{
+			// XXX: We need an actual converter here.
+			Converter: failConverter{},
+		}
+		// XXX: Currently using the first type, since it's not
+		// easy to find the type we care about?
+		p.mechanism = &applyPatcher{patcher: p, converter: converter, updater: updater}
+		p.applyManagedFields = noOpTransform
 		p.forceAllowCreate = true
 	default:
 		return nil, false, fmt.Errorf("%v: unimplemented patch type", p.patchType)
 	}
 
 	wasCreated := false
-	p.updatedObjectInfo = rest.DefaultUpdatedObjectInfo(nil, p.applyPatch, p.applyAdmission)
+	p.updatedObjectInfo = rest.DefaultUpdatedObjectInfo(nil, p.applyPatch, p.applyManagedFields, p.applyAdmission)
```

Comment:
You're passing nil here, which is supposed to be `newObject`. My problem with using newObject for the patch is that the order matters  a lot here :-/

Reply:
So this is fixed with using patchBytes again?

Reply:
yes